Generate .sym file

Lapixx opened the pull request:
Generates a sym file for all labels. Checked that it works on bgb as well.

Lapixx commented on dmgforth:
```diff
@@ -55,4 +56,5 @@ depth 0<> [if]
 fix-header-complement
 fix-global-checksum
 output-file dump-rom
+s" output.sym" dump-sym
```
This should probably be `<output-file>.sym`, but didn't know how to properly do this.

davazp replied:
String manipulation is pretty tricky in Forth. You can use `pad`, which just pushes the address of a temporary region of memory useful to do string manipulation.

Doing so, I got this working:

```forth
: append-to { c-addr u addr -- addr' }
  c-addr addr u move
  addr u + ;
  
: append-string ( c-addr1 u1 c-addr2 u2 -- c-addr3 u3 )
  2swap dup >r
  pad append-to

  over >r
  append-to

  drop pad r> r> + ;

( Remove the extension of a filename )
: sans-extname ( c-addr u -- c-addr' u' )
  dup 1+ 1 ?do
    2dup i - + c@
    [char] . = if i - unloop exit then
  loop ;

: symbol-file ( c-addr u -- c-addr' u' )
  sans-extname s" .sym" append-string ;

: output-file ( c-addr u -- c-addr' u' )
  sans-extname s" .gb" append-string ;

```

then you can use it:

```
s" input.fs" symbol-file  type ( prints input.sym )
```

Note that pad is a transient memory. The next pad will invalidate it. So you can't store the resulting string in a variable or anything.

You must use it immediately, like in

`input symbol-file create-file`

davazp replied:
If you want persistent memory, you can use dynamic memory allocation

https://www.complang.tuwien.ac.at/forth/gforth/Docs-html/Heap-Allocation.html#Heap-Allocation

but then it needs to be free-ed later on.  Or you can just use dictionary allocation if you know it won't be a problem by using up all the memory.

davazp commented on src/sym.fs:
```diff
@@ -0,0 +1,58 @@
+require ./utils.fs
+
+32 kB constant sym-size
```
For example, what about

```forth
s" output.sym" w/o create-file throw Value sym-out

: sym-cr
  #10 sym-out emit-file throw ;

: sym-write ( c-addr u --)
  sym-out write-file throw
  sym-out flush-file throw  ;

: addr-to-str ( nn -- c-addr u )
  base @ >r
  hex
  0
  <<# # # # # #> #>>
  r> BASE ! ;

: sym-emit ( c-addr u nn -- )
  s" 00:"       sym-write
  addr-to-str   sym-write
  s"  "         sym-write
  ( c-addr u )  sym-write
  sym-cr ;

s" ; Generated by dmg-forth " sym-write sym-cr
s" symbol-1" $ff00 sym-emit
s" symbol-2" $ff20 sym-emit
s" symbol-3" $efef sym-emit
```
?

Lapixx commented:
Simplified the `sym.fs` file with your suggested changes. Also added the filename generation for the .sym file. Left the output-file as it is for now (from CLI), but we could probably extend it a bit in the future to convert `input.fs` -> `input.gb` if no output name is specified.
